File: benchmarks/py-fastapi-tenant-leak/tests_hidden/grader.py
import sys

sys.path.insert(0, "/workspace")
from app import Invoice, InvoiceRepository, User, list_open_invoices

mode = sys.argv[1]


def ids(tenant: str, invoices: list[Invoice]) -> list[int]:
    return [item.id for item in list_open_invoices(User("u", tenant), InvoiceRepository(invoices))]


if mode == "hidden":
    rows = [Invoice(1, "acme", "open"), Invoice(2, "other", "open"), Invoice(3, "acme", "closed")]
    assert ids("acme", rows) == [1]
elif mode == "security":
    rows = [Invoice(1, "t", "open"), Invoice(2, "t-admin", "open"), Invoice(3, "t", "open")]
    assert ids("t", rows) == [1, 3]
elif mode == "mutation":
    rows = [Invoice(i, f"tenant-{i % 5}", "open") for i in range(50)]
    assert ids("tenant-3", rows) == [3, 8, 13, 18, 23, 28, 33, 38, 43, 48]
elif mode == "regression":
    rows = [Invoice(1, "same", "closed"), Invoice(2, "same", "open")]
    assert ids("same", rows) == [2]
else:
    raise SystemExit(2)

File: benchmarks/ts-tenant-authorization/tests_hidden/grader.ts
import assert from "node:assert/strict";
import { canEditProject } from "/workspace/src/authorization.ts";

const mode = process.argv[2];
const project = { id: "p", tenantId: "t1", ownerId: "owner" } as const;

if (mode === "hidden") {
  assert.equal(canEditProject({ id: "x", tenantId: "t1", role: "admin" }, project), true);
  assert.equal(canEditProject({ id: "owner", tenantId: "t1", role: "member" }, project), true);
  assert.equal(canEditProject({ id: "x", tenantId: "t1", role: "member" }, project), false);
} else if (mode === "security") {
  assert.equal(canEditProject({ id: "owner", tenantId: "t2", role: "member" }, project), false);
  assert.equal(canEditProject({ id: "x", tenantId: "t2", role: "admin" }, project), false);
} else if (mode === "mutation") {
  const tricky = { id: "p", tenantId: "tenant", ownerId: "u" } as const;
  assert.equal(canEditProject({ id: "u", tenantId: "tenant-admin", role: "admin" }, tricky), false);
} else if (mode === "regression") {
  assert.equal(canEditProject({ id: "same", tenantId: "t1", role: "admin" }, project), true);
} else {
  process.exit(2);
}
